src/System.Data.Common/src/System/Data/Common/DbConnectionOptions.cs

```diff
@@ -0,0 +1,621 @@
+                  // embedded within a value, the retail code does.
+                  // +  "(?<![\\}]))"                                            // unquoted value must not stop with }
+
+            + ")(\\s*)(;|[\u0000\\s]*$)"                               // whitespace after value up to semicolon or end-of-line
+            + ")*"                                                      // repeat the key-value pair
+            + "[\\s;]*[\u0000\\s]*"                                     // traling whitespace/semicolons (DataSourceLocator), embedded nulls are allowed only in the end
+        ;
+
+        private static readonly Regex s_connectionStringRegex = new Regex(ConnectionStringPattern, RegexOptions.ExplicitCapture);
+#endif
+        private const string ConnectionStringValidKeyPattern = "^(?![;\\s])[^\\p{Cc}]+(?<!\\s)$"; // key not allowed to start with semi-colon or space or contain non-visible characters or end with space
+        private const string ConnectionStringValidValuePattern = "^[^\u0000]*$";                    // value not allowed to contain embedded null
+        private const string ConnectionStringQuoteValuePattern = "^[^\"'=;\\s\\p{Cc}]*$";           // generally do not quote the value if it matches the pattern
+        private const string ConnectionStringQuoteOdbcValuePattern = "^\\{([^\\}\u0000]|\\}\\})*\\}$"; // do not quote odbc value if it matches this pattern
+        internal const string DataDirectory = "|datadirectory|";
+
+        private static readonly Regex s_connectionStringValidKeyRegex = new Regex(ConnectionStringValidKeyPattern);
+        private static readonly Regex s_connectionStringValidValueRegex = new Regex(ConnectionStringValidValuePattern);
+        private static readonly Regex s_connectionStringQuoteValueRegex = new Regex(ConnectionStringQuoteValuePattern);
+
+        // connection string common keywords
+        private static class KEY
+        {
+            internal const string Integrated_Security = "integrated security";
+            internal const string Password = "password";
+            internal const string Persist_Security_Info = "persist security info";
+            internal const string User_ID = "user id";
+        };
+
+        // known connection string common synonyms
+        private static class SYNONYM
+        {
+            internal const string Pwd = "pwd";
+            internal const string UID = "uid";
+        };
+
+        private readonly string _usersConnectionString;
+        private readonly Hashtable _parsetable;
+        internal readonly NameValuePair KeyChain;
+        internal readonly bool HasPasswordKeyword;
+
+
+        // synonyms hashtable is meant to be read-only translation of parsed string
+        // keywords/synonyms to a known keyword string
+        public DbConnectionOptions(string connectionString, Hashtable synonyms)
+        {
+            _parsetable = new Hashtable();
```

Comment:
_parsetable only appears to be used in the constructor.  Does it need to be a field?  And if so, could it be a strongly-typed ```Dictionary<TKey, TValue>``` instead of the legacy, weakly-typed ```Hashtable```?